app/controllers/msb/WhichCurrenciesController.scala

```diff
@@ -50,21 +50,31 @@ class WhichCurrenciesController @Inject() (val authConnector: AuthConnector,
     implicit authContext => implicit request => {
       Form2[WhichCurrencies](request.body) match {
         case f: InvalidForm =>
-          Future.successful(BadRequest(views.html.msb.which_currencies(f, edit)))
+          Future.successful(BadRequest(views.html.msb.which_currencies(removeEmptyFields(f), edit)))
         case ValidForm(_, data: WhichCurrencies) =>
               for {
                 msb <- dataCacheConnector.fetch[MoneyServiceBusiness](MoneyServiceBusiness.key)
                 _ <- dataCacheConnector.save[MoneyServiceBusiness](MoneyServiceBusiness.key,
                   updateCurrencies(msb, data))
-              } yield
-                  edit match {
-                    case true => Redirect(routes.SummaryController.get())
-                    case _ => Redirect(routes.UsesForeignCurrenciesController.get())
-                  }
+              } yield edit match {
+                case true => Redirect(routes.SummaryController.get())
+                case _ => Redirect(routes.UsesForeignCurrenciesController.get())
+              }
       }
     }
   }
 
+  // Validation removes empty fields. This removes them from the form data, so the errors align correctly with the fields.
```

Comment:
Not sure we need the comment here. Goes against centre guidelines, Code should be explanatory when read which it is. other than that sound job :+1:

Reply:
I disagree about comments, I think they can be useful for many reasons - for example, providing context. Reading code provides what and how, but it might not be self-evident why. 

If a comment saves someone valuable time wondering why a piece of code is there, removing it, breaking things, having to re-add it (potentially in a few years, after every current dev has left), then I think they're valid.

That being said, if there are centre guidelines you can point me to that say "DO NOT USE COMMENTS", then I can remove it. But I still disagree :)

Reply:
We’ve discussed the use of comments multiple times in the past and came to the agreement not to use them. It is centre guidelines I’m not sure where it’s written but I’ve been told by nick wolfie etc not to do it. If you have a big issue removing it maybe ask The rest of the team about this

Reply:
I don't have a big issue over this particular comment. We've discussed comments before, but I don't think the issue has ever been definitely settled, as I said, there are valid reasons for using them. If it is a centre guideline, I'd like to see the guidelines, if only to know how else I might inadvertently be contravening the centre's house-style.

Reply:
Thanks for removing :+1: